src/DotVVM.Framework/Compilation/Parser/Binding/Parser/TypeDeclarationBindingParserNode.cs

```diff
@@ -0,0 +1,78 @@
+﻿#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DotVVM.Framework.Compilation.Parser.Binding.Parser
+{
+    public enum TypeModifier
+    {
+        None,
+        Array,
+        Nullable,        
+    }
+
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public class TypeDeclarationBindingParserNode : BindingParserNode
```

Comment:
This types seems overloaded - it should be abstract without the Type and Next properties as these are only used in certain subtypes.

Also, I'd call this a type reference - definition would be a `class X { }` syntax.

I'd imagine hierarchy likes this:

```csharp

abstract class TypeReferenceBindingParserNode
sealed class ArrayTypeReferenceBindingParserNode(TypeReferenceBindingParserNode elementType) : TypeReferenceBindingParserNode
sealed class NullableTypeReferenceBindingParserNode(TypeReferenceBindingParserNode innerType) : TypeReferenceBindingParserNode
sealed class ActualTypeReferenceBindingParserNode(BindingParserNode type) : TypeReferenceBindingParserNode
```

Including generic types into this hierarchy would be nice.